File: fullstack/auth/paseto-cross-language/issuer-go/internal/protected/service.go
// Package protected is the Go implementation of demo.v1.ProtectedService —
// the same service Rust and TS implement, verifying the same tokens with the
// same public key.
package protected

import (
	"context"
	"errors"
	"strings"
	"time"

	"connectrpc.com/connect"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/reflect/protoreflect"
	"google.golang.org/protobuf/reflect/protoregistry"
	"google.golang.org/protobuf/types/descriptorpb"

	"github.com/Skayfa/reference-patterns/fullstack/auth/paseto-cross-language/issuer-go/internal/token"
	authv1 "github.com/Skayfa/reference-patterns/fullstack/auth/paseto-cross-language/issuer-go/pb/auth/v1"
	demov1 "github.com/Skayfa/reference-patterns/fullstack/auth/paseto-cross-language/issuer-go/pb/demo/v1"
)

type claimsKey struct{}

// AccessRuleFor reads the (auth.v1.access) option declared on the RPC in
// the proto — the contract itself says who may call it. nil = the RPC was
// never annotated, which the default-deny posture treats as forbidden.
func AccessRuleFor(procedure string) *authv1.AccessRule {
	name := strings.ReplaceAll(strings.TrimPrefix(procedure, "/"), "/", ".")
	desc, err := protoregistry.GlobalFiles.FindDescriptorByName(protoreflect.FullName(name))
	if err != nil {
		return nil
	}
	method, ok := desc.(protoreflect.MethodDescriptor)
	if !ok {
		return nil
	}
	opts, ok := method.Options().(*descriptorpb.MethodOptions)
	if !ok {
		return nil
	}
	rule, _ := proto.GetExtension(opts, authv1.E_Access).(*authv1.AccessRule)
	if rule.GetMinimumRole() == authv1.Role_ROLE_UNSPECIFIED && !rule.GetPublic() {
		return nil
	}
	return rule
}

// RoleLevel maps a token's role claim ("admin") onto the auth.v1.Role
// hierarchy (ROLE_ADMIN). Unknown claims map to ROLE_UNSPECIFIED and pass
// nothing.
func RoleLevel(claim string) authv1.Role {
	return authv1.Role(authv1.Role_value["ROLE_"+strings.ToUpper(claim)])
}

// NewAuthInterceptor guards the authenticated services: it always requires a
// valid Bearer PASETO, enforces the minimum role the proto declares for the
// RPC (default-deny when no rule is declared), and stashes the claims in the
// context. Verification is local: only the public key.
//
// Truly public RPCs (public: true — the whole AuthService) are mounted on a
// separate handler WITHOUT this interceptor; see newMux. This interceptor is
// never placed in front of a public RPC, so it always demands a token. A
// public rule reaching it is therefore a mount mistake, and is refused rather
// than silently waved through.
func NewAuthInterceptor(verifier *token.Verifier) connect.UnaryInterceptorFunc {
	return func(next connect.UnaryFunc) connect.UnaryFunc {
		return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
			raw, ok := strings.CutPrefix(req.Header().Get("Authorization"), "Bearer ")
			if !ok || raw == "" {
				return nil, connect.NewError(connect.CodeUnauthenticated,
					errors.New("missing bearer token"))
			}
			claims, err := verifier.Verify(raw)
			if err != nil {
				return nil, connect.NewError(connect.CodeUnauthenticated,
					errors.New("invalid token"))
			}
			rule := AccessRuleFor(req.Spec().Procedure)
			switch {
			case rule == nil:
				return nil, connect.NewError(connect.CodePermissionDenied,
					errors.New("no access rule declared for this rpc"))
			case rule.GetPublic():
				return nil, connect.NewError(connect.CodePermissionDenied,
					errors.New("public rpc mounted behind the auth interceptor — mount it without one"))
			case RoleLevel(claims.Role) < rule.GetMinimumRole():
				return nil, connect.NewError(connect.CodePermissionDenied,
					errors.New(roleName(rule.GetMinimumRole())+" role required"))
			}
			return next(context.WithValue(ctx, claimsKey{}, claims), req)
		}
	}
}

func roleName(role authv1.Role) string {
	return strings.ToLower(strings.TrimPrefix(role.String(), "ROLE_"))
}

func ClaimsFrom(ctx context.Context) (token.Claims, error) {
	claims, ok := ctx.Value(claimsKey{}).(token.Claims)
	if !ok {
		return token.Claims{}, connect.NewError(connect.CodeInternal,
			errors.New("handler mounted without auth interceptor"))
	}
	return claims, nil
}

type Service struct{}

func (Service) WhoAmI(
	ctx context.Context, _ *connect.Request[demov1.WhoAmIRequest],
) (*connect.Response[demov1.WhoAmIResponse], error) {
	claims, err := ClaimsFrom(ctx)
	if err != nil {
		return nil, err
	}
	return connect.NewResponse(&demov1.WhoAmIResponse{
		Subject:   claims.Subject,
		Role:      claims.Role,
		IssuedAt:  claims.IssuedAt.Format(time.RFC3339),
		ExpiresAt: claims.ExpiresAt.Format(time.RFC3339),
		ServedBy:  "go-connect",
	}), nil
}

func (Service) AdminOnly(
	ctx context.Context, _ *connect.Request[demov1.AdminOnlyRequest],
) (*connect.Response[demov1.AdminOnlyResponse], error) {
	// The role check happened in the interceptor, driven by the proto's
	// (auth.v1.access) option — nothing to re-check here.
	if _, err := ClaimsFrom(ctx); err != nil {
		return nil, err
	}
	return connect.NewResponse(&demov1.AdminOnlyResponse{
		Secret: "the go server trusts your admin token",
	}), nil
}

File: fullstack/auth/paseto-cross-language/verifier-rust/src/service.rs
//! demo.v1.ProtectedService in Rust: the same service Go and TS implement,
//! verifying the same tokens with the same public key.

use std::sync::Arc;

use tonic::{Request, Response, Status};

use crate::access::{AccessRules, authorize};
use crate::paseto::{Claims, Verifier};
use crate::pb::demo_v1::protected_service_server::ProtectedService;
use crate::pb::demo_v1::{
    AdminOnlyRequest, AdminOnlyResponse, WhoAmIRequest, WhoAmIResponse,
};

pub const SERVED_BY: &str = "rust-tonic";

/// Verifies the Bearer PASETO on every request and stashes the claims in the
/// request extensions. Verification is local: only the public key.
#[derive(Clone)]
pub struct AuthInterceptor {
    verifier: Arc<Verifier>,
}

impl AuthInterceptor {
    pub fn new(verifier: Verifier) -> Self {
        Self { verifier: Arc::new(verifier) }
    }
}

impl tonic::service::Interceptor for AuthInterceptor {
    fn call(&mut self, mut request: Request<()>) -> Result<Request<()>, Status> {
        let header = request
            .metadata()
            .get("authorization")
            .and_then(|v| v.to_str().ok())
            .unwrap_or_default();
        let raw = header
            .strip_prefix("Bearer ")
            .filter(|t| !t.is_empty())
            .ok_or_else(|| Status::unauthenticated("missing bearer token"))?;
        let claims = self
            .verifier
            .verify(raw)
            .map_err(|_| Status::unauthenticated("invalid token"))?;
        request.extensions_mut().insert(claims);
        Ok(request)
    }
}

pub struct ProtectedServiceImpl {
    rules: AccessRules,
}

impl ProtectedServiceImpl {
    pub fn new() -> Self {
        Self { rules: AccessRules::from_contract() }
    }
}

#[tonic::async_trait]
impl ProtectedService for ProtectedServiceImpl {
    async fn who_am_i(
        &self,
        request: Request<WhoAmIRequest>,
    ) -> Result<Response<WhoAmIResponse>, Status> {
        let claims = authorize(&self.rules, &request, "demo.v1.ProtectedService.WhoAmI")?;
        Ok(Response::new(WhoAmIResponse {
            subject: claims.subject,
            role: claims.role,
            issued_at: claims.issued_at,
            expires_at: claims.expires_at,
            served_by: SERVED_BY.to_owned(),
        }))
    }

    async fn admin_only(
        &self,
        request: Request<AdminOnlyRequest>,
    ) -> Result<Response<AdminOnlyResponse>, Status> {
        authorize(&self.rules, &request, "demo.v1.ProtectedService.AdminOnly")?;
        Ok(Response::new(AdminOnlyResponse {
            secret: "the rust server trusts your admin token".to_owned(),
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::paseto::DEV_PUBLIC_HEX;
    use tonic::metadata::MetadataValue;
    use tonic::service::Interceptor;

    const GO_SIGNED_ADMIN: &str = include_str!("../../keys/go-signed-admin.paseto");

    fn intercepted(token: Option<&str>) -> Result<Request<()>, Status> {
        let mut request = Request::new(());
        if let Some(token) = token {
            request.metadata_mut().insert(
                "authorization",
                MetadataValue::try_from(format!("Bearer {token}")).unwrap(),
            );
        }
        AuthInterceptor::new(Verifier::from_hex(DEV_PUBLIC_HEX).unwrap()).call(request)
    }

    fn with_claims<T>(message: T, role: &str) -> Request<T> {
        let mut request = Request::new(message);
        request.extensions_mut().insert(Claims {
            subject: "user-1".to_owned(),
            role: role.to_owned(),
            issued_at: "2026-07-15T12:00:00Z".to_owned(),
            expires_at: "2036-07-12T12:00:00Z".to_owned(),
        });
        request
    }

    #[test]
    fn interceptor_accepts_go_signed_token_and_stashes_claims() {
        let request = intercepted(Some(GO_SIGNED_ADMIN)).unwrap();
        let claims = request.extensions().get::<Claims>().unwrap();
        assert_eq!(claims.subject, "fixture-admin");
        assert_eq!(claims.role, "admin");
    }

    #[test]
    fn interceptor_rejects_missing_and_invalid_tokens() {
        assert_eq!(intercepted(None).unwrap_err().code(), tonic::Code::Unauthenticated);
        assert_eq!(intercepted(Some("")).is_err(), true);
        let tampered = GO_SIGNED_ADMIN.replace('a', "b");
        assert_eq!(
            intercepted(Some(&tampered)).unwrap_err().code(),
            tonic::Code::Unauthenticated
        );
    }

    #[tokio::test]
    async fn who_am_i_echoes_claims() {
        let response = ProtectedServiceImpl::new()
            .who_am_i(with_claims(WhoAmIRequest {}, "user"))
            .await
            .unwrap();
        let msg = response.into_inner();
        assert_eq!(msg.subject, "user-1");
        assert_eq!(msg.served_by, SERVED_BY);
    }

    #[tokio::test]
    async fn admin_only_enforces_role() {
        let denied = ProtectedServiceImpl::new()
            .admin_only(with_claims(AdminOnlyRequest {}, "user"))
            .await
            .unwrap_err();
        assert_eq!(denied.code(), tonic::Code::PermissionDenied);

        ProtectedServiceImpl::new()
            .admin_only(with_claims(AdminOnlyRequest {}, "admin"))
            .await
            .unwrap();
    }

    #[tokio::test]
    async fn handler_without_interceptor_is_an_internal_error() {
        let err = ProtectedServiceImpl::new()
            .who_am_i(Request::new(WhoAmIRequest {}))
            .await
            .unwrap_err();
        assert_eq!(err.code(), tonic::Code::Internal);
    }
}
